pkg/server/api/agent/v1/service.go

```diff
@@ -35,7 +38,65 @@ type Service struct {
 }
 
 func (s *Service) ListAgents(ctx context.Context, req *agent.ListAgentsRequest) (*agent.ListAgentsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "method not implemented")
+	log := rpccontext.Logger(ctx)
+
+	listReq := &datastore.ListAttestedNodesRequest{}
+
+	if req.OutputMask == nil || req.OutputMask.Selectors {
+		listReq.FetchSelectors = true
+	}
+	// Parse proto filter into datastore request
+	if req.Filter != nil {
+		filter := req.Filter
+		listReq.ByAttestationType = filter.ByAttestationType
+		listReq.ByBanned = filter.ByBanned
+
+		if filter.BySelectorMatch != nil {
+			selectors, err := api.SelectorsFromProto(filter.BySelectorMatch.Selectors)
+			if err != nil {
+				log.WithError(err).Error("Failed to parse selectors")
+				return nil, status.Errorf(codes.InvalidArgument, "failed to parse selectors: %v", err)
+			}
+			listReq.BySelectorMatch = &datastore.BySelectors{
+				Match:     datastore.BySelectors_MatchBehavior(filter.BySelectorMatch.Match),
+				Selectors: selectors,
+			}
+		}
+	}
+
+	// Set pagination parameters
+	if req.PageSize > 0 {
+		listReq.Pagination = &datastore.Pagination{
+			PageSize: req.PageSize,
+			Token:    req.PageToken,
+		}
+	}
+
+	dsResp, err := s.ds.ListAttestedNodes(ctx, listReq)
+	if err != nil {
+		log.WithError(err).Error("Failed to list agents")
+		return nil, status.Errorf(codes.Internal, "failed to list agents: %v", err)
+	}
+
+	resp := &agent.ListAgentsResponse{}
+
+	if dsResp.Pagination != nil {
+		resp.NextPageToken = dsResp.Pagination.Token
+	}
+
+	// Parse nodes into proto and apply output mask
+	for _, node := range dsResp.Nodes {
+		a, err := api.ProtoFromAttestedNode(node)
+		if err != nil {
+			log.WithError(err).WithField(telemetry.SPIFFEID, node.SpiffeId).Error("Unable to parse attested node")
```

Comment:
I'm more inclined to log this at WARN level, since we don't fail the call and we continue processing the results.